File: 36.有效的数独.java
/*
 * @lc app=leetcode.cn id=36 lang=java
 *
 * [36] 有效的数独
 */

// @lc code=start
class Solution {
    public boolean isValidSudoku(char[][] board) {
        int[][] rowFreq = new int[9][9];
        int[][] colFreq = new int[9][9];
        int[][] subboxFreq = new int[9][9];

        for (int i = 0; i < 9; ++i) {
            for (int j = 0; j < 9; ++j) {
                char c = board[i][j];
                if (c == '.')  continue;
                int idx = c - '0' - 1;
                int k = 3 * (i / 3) + j / 3;
                ++rowFreq[i][idx];
                ++colFreq[j][idx];
                ++subboxFreq[k][idx];
                if (rowFreq[i][idx] > 1 || colFreq[j][idx] > 1 || subboxFreq[k][idx] > 1)
                    return false;
            }
        }
        return true;
    }
}
// @lc code=end

File: 36.有效的数独.py
#
# @lc app=leetcode.cn id=36 lang=python3
#
# [36] 有效的数独
#
from typing import List
from pprint import pprint

# @lc code=start
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        row_num_counts = [[0]*9 for _ in range(9)]
        col_num_counts = [[0]*9 for _ in range(9)]
        subgrid_num_counts = [[0]*9 for _ in range(9)]
        for i in range(9):
            for j in range(9):
                num = board[i][j]   # or [j][i]
                if num == '.':
                    continue
                num = int(num)
                num_idx = num-1
                subgrid_idx = 3*(i//3)+j//3
                row_num_counts[i][num_idx] += 1
                col_num_counts[j][num_idx] += 1
                subgrid_num_counts[subgrid_idx][num_idx] += 1
                if row_num_counts[i][num_idx] > 1 or \
                    col_num_counts[j][num_idx] > 1 or \
                    subgrid_num_counts[subgrid_idx][num_idx] > 1:
                        return False
        return True
        

# @lc code=end
